### PR TITLE
Sharapov Georgiy. First Lab.

### PR DESCRIPTION
**Pull request by nol0n**
Added plugin that emits a warning for each function that has the deprecated attribute.

Plugin has two arguments

An option `--attr-style` that determines which style of attributes will trigger a warning. Possible values are `gcc`, `c++14`, and `both`. The default value is set to `both`.

A `-help` flag that displays information about `--attr-style`.

### CODE REVIEW

File: clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/FrontendAction.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+class DepWarningVisitor : public clang::RecursiveASTVisitor<DepWarningVisitor> {
+ public:
```
Comment (m-ly4) [inline clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp]: Could you please align your code with the help of `clang-format` tool?

File: clang/test/lab1/sharapov_georgiy/test.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// RUN: %clang++ -cc1 -load %llvmshlibdir/DepWarningPlugin%pluginext -plugin DepEmitWarning %s 2>&1 | FileCheck %s
+
+// CHECK: test.cpp:5:6: warning: function 'oldSum' is deprecated
+[[deprecated]]
+void oldSum(int a, int b) {}
+
+void newSum(int a, int b) {}
```
Comment (m-ly4) [inline clang/test/lab1/sharapov_georgiy/test.cpp]: Could you please check that there is no `deprecated` warning?

File: clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    bool VisitFunctionDecl(clang::FunctionDecl *F) {
+        if (F->hasAttr<clang::DeprecatedAttr>()) {
```
Comment (m-ly4) [inline clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp]: Could you please also cover gcc-style `deprecated` attribute?

File: clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+class DepWarningAction : public clang::PluginASTAction {
+ public:
+    std::unique_ptr<clang::ASTConsumer> CreateASTConsumer(
+        clang::CompilerInstance &Compiler, llvm::StringRef InFile) override {
+        return std::make_unique<DepWarningConsumer>();
+    }
+ protected:
+    bool ParseArgs(const clang::CompilerInstance &Compiler, const std::vector<std::string> &args) override {
+        return true;
```
Comment (m-ly4) [inline clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp]: Could you please add `attr-style` argument specifying which style of `deprecated` attribute to cover? Options may be `[c++14, gcc, both]`.
Please also add tests for these arguments.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/sharapov_georgiy/test.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// CHECK: wrong_style.cpp:13:6: warning: function 'oldFunc3' is deprecated (gcc)
+__attribute__((deprecated))
+void oldFunc3(int a, int b);
+
+// CHECK: wrong_style.cpp:17:6: warning: function 'oldFunc4' is deprecated (c++14)
+[[deprecated]]
+void oldFunc4(int a, int b);
+
+// CHECK-NOT: warning:
+void Func(int a, int b);
```
Comment (m-ly4) [inline clang/test/lab1/sharapov_georgiy/test.cpp]:
```suggestion
void Func(int a, int b);

```

File: clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                        "attributes will be displayed,"
+                        ""
+                        ""
+                        "\n                             options: gcc, c++14, "
+                        "both. Default value 'both'.\n\n"
+                        "";
+      } else if (Arg.find("--attr-style=") == 0) {
+        attrStyle = Arg.substr(std::string("--attr-style=").length());
+        if (attrStyle != "c++14" && attrStyle != "gcc" && attrStyle != "both") {
+          llvm::outs() << "\n  Invalid value for attr-style. Using default "
```
Comment (m-ly4) [inline clang/labs/lab1/sharapov_georgiy/DepWarningPlugin.cpp]: Could you please report unexpected option parameter warning via diagnostic?
And also check it in a test?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by nol0n** [comment]
@m-ly4, could you please take a look at the windows-build? This commit in my fork passed tests without issues ([link to the test results](https://github.com/nol0n/llvm-nnsu-2024/actions/runs/8276739737/job/22645817184)) and as far as I can tell, the problem doesn't seem to be with my code. However, I might be mistaken.

**Comment by m-ly4** [comment]
> @m-ly4, could you please take a look at the windows-build? This commit in my fork passed tests without issues ([link to the test results](https://github.com/nol0n/llvm-nnsu-2024/actions/runs/8276739737/job/22645817184)) and as far as I can tell, the problem doesn't seem to be with my code. However, I might be mistaken.

Looks like it was just infrastructure problem. I restarted windows test.

**Review by m-ly4 — APPROVED** [review]

**Comment by aobolensk** [comment]
LGTM

**Review by aobolensk — APPROVED** [review]

**Pull request merged**